portable/NetworkInterface/TM4C/NetworkInterface.c

```diff
@@ -0,0 +1,599 @@
+        // packets along with those addressed specifically for us.
+        MAP_EMACFrameFilterSet(EMAC0_BASE, (EMAC_FRMFILTER_HASH_AND_PERFECT |
+                           EMAC_FRMFILTER_PASS_MULTICAST));
+
+        // Set the MAC address
+        MAP_EMACAddrSet(EMAC0_BASE, 0, &mac_address_bytes[0]);
+
+        // Clears any previously asserted interrupts
+        MAP_EMACIntClear(EMAC0_BASE, EMACIntStatus(EMAC0_BASE, false));
+
+        // Initialize the DMA descriptors
+        _dma_descriptors_init();
+
+        // Enable TX/RX
+        MAP_EMACTxEnable(EMAC0_BASE);
+        MAP_EMACRxEnable(EMAC0_BASE);
+
+        // Set the interrupt to a lower priority than the OS scheduler interrupts
+        MAP_IntPrioritySet(INT_EMAC0,  (6 << (8 - configPRIO_BITS)));
+
+        // Enable the Ethernet RX and TX interrupt source.
+        MAP_EMACIntEnable(EMAC0_BASE, (EMAC_INT_RECEIVE | EMAC_INT_TRANSMIT |
+                      EMAC_INT_TX_STOPPED | EMAC_INT_RX_NO_BUFFER |
+                      EMAC_INT_RX_STOPPED | EMAC_INT_PHY));
+
+        // Enable EMAC interrupts
+        MAP_IntEnable(INT_EMAC0);
+
+        // Enable all processor interrupts.
+        MAP_IntMasterEnable();
+
+        // Tell the PHY to start an auto-negotiation cycle.
+        MAP_EMACPHYWrite(EMAC0_BASE, PHY_PHYS_ADDR, EPHY_BMCR, (EPHY_BMCR_ANEN |
+                EPHY_BMCR_RESTARTAN));
+    }
+
+    return success;
+}
+
+BaseType_t xNetworkInterfaceOutput(NetworkBufferDescriptor_t * const pxNetworkBuffer, BaseType_t xReleaseAfterSend)
+{
+    BaseType_t success = pdTRUE;
+    tEMACDMADescriptor *dma_descriptor;
+
+    dma_descriptor = &_tx_descriptors[_tx_descriptor_list.write];
+
+    // If the DMA controller still owns the descriptor, all DMA descriptors are in use, bail out
+    if (0 == (dma_descriptor->ui32CtrlStatus & DES0_RX_CTRL_OWN))
+    {
+        // Assign the buffer to the DMA descriptor
+        dma_descriptor->pvBuffer1 = pxNetworkBuffer->pucEthernetBuffer;
+
+        // Inform the DMA of the size of the packet
+        dma_descriptor->ui32Count = (pxNetworkBuffer->xDataLength & DES1_TX_CTRL_BUFF1_SIZE_M) << DES1_TX_CTRL_BUFF1_SIZE_S;
+
+        // Inform the DMA that this is the first and last segment of the packet, calculate the checksums, the descriptors are
+        // chained, and to use interrupts
+        dma_descriptor->ui32CtrlStatus = DES0_TX_CTRL_FIRST_SEG | /*DES0_TX_CTRL_IP_ALL_CKHSUMS |*/ DES0_TX_CTRL_CHAINED
+                | DES0_TX_CTRL_LAST_SEG | DES0_TX_CTRL_INTERRUPT;// | DES0_TX_CTRL_REPLACE_CRC;
+
+        // Advance the index in the list
+        _tx_descriptor_list.write++;
+
+        // Wrap around if required
+        if (_tx_descriptor_list.write == SYSCONFIG_FREERTOS_TCP_TX_DESC_COUNT)
+        {
+            _tx_descriptor_list.write = 0;
+        }
+
+        // Give the DMA descriptor to the DMA controller
+        dma_descriptor->ui32CtrlStatus |= DES0_TX_CTRL_OWN;
```

Comment:
Did this work properly in your hardware? Often it is necessary to create a memory barrier, e.g. by reading back the descriptor.
Is D-caching disabled for the DMA-descriptor memory?

Reply:
This does work in hardware and is how TI implemented their reference driver for lwIP. I'll see if I can find references in the datasheet to confirm.

Reply:
I can find no information regarding a memory barrier being needed here, and the TI supplied lwIP driver does not use volatile keywords, etc. to enforce a barrier. The DMA apparently fetches the TX descriptors when a transmission is requested:

![TXDescriptorTDES0](https://user-images.githubusercontent.com/94705449/155209582-2b18d960-8007-4740-b093-c51d27e6c2dc.png)

The RX descriptor documentation only mentions that this bit is cleared.

With that in mind, and test code working as expected, I feel comfortable using this bit as is.